metric-anomaly-detector/src/main/java/org/hypertrace/alert/engine/metric/anomaly/detector/evaluator/MetricCache.java

```diff
@@ -0,0 +1,141 @@
+      String tenantId,
+      long startTimeMillis,
+      long endTimeMillis) {
+    long metricDurationMillis = endTimeMillis - startTimeMillis;
+
+    MetricTimeSeries metricTimeSeries = metricCache.getIfPresent(metricSelection);
+    // no record for this metric selection
+    // if requested time range is earlier than existing
+    // replace data
+    if (null == metricTimeSeries || startTimeMillis < metricTimeSeries.getStartTimeMillis()) {
+      Iterator<ResultSetChunk> iterator =
+          queryRequestHandler.executeQuery(
+              requestHeaders, metricSelection, tenantId, startTimeMillis, endTimeMillis);
+      List<Pair<Long, Double>> dataList = consumeIterator(iterator);
+      metricCache.put(
+          metricSelection,
+          new MetricTimeSeries(startTimeMillis, endTimeMillis, dataList, metricDurationMillis));
+      return dataList;
+    }
+
+    // need to just get diff of the data
+    if (endTimeMillis > metricTimeSeries.getEndTimeMillis()) {
+      Iterator<ResultSetChunk> iterator =
+          queryRequestHandler.executeQuery(
+              requestHeaders,
+              metricSelection,
+              tenantId,
+              metricTimeSeries.getEndTimeMillis(),
+              endTimeMillis);
+      List<Pair<Long, Double>> dataList = consumeIterator(iterator);
+      metricTimeSeries.getDataList().addAll(dataList);
+      metricTimeSeries.setEndTimeMillis(endTimeMillis);
+      metricTimeSeries.setMaxRetentionPeriodMillis(
+          Math.max(metricTimeSeries.getMaxRetentionPeriodMillis(), metricDurationMillis));
+
+      // trim off any stale data
+      metricTimeSeries.trimOlderData();
+    }
+
+    return filterData(metricTimeSeries.getDataList(), startTimeMillis, endTimeMillis);
+  }
+
+  private List<Pair<Long, Double>> consumeIterator(Iterator<ResultSetChunk> iterator) {
+    List<Pair<Long, Double>> list = new ArrayList<>();
+    while (iterator.hasNext()) {
+      ResultSetChunk resultSetChunk = iterator.next();
+      for (Row row : resultSetChunk.getRowList()) {
+        Value value = row.getColumn(1);
+        if (value.getValueType() != ValueType.STRING) {
+          throw new IllegalArgumentException(
+              "Expecting value of type string, received valueType: " + value.getValueType());
+        }
+        list.add(
+            Pair.of(
+                Long.parseLong(row.getColumn(0).getString()),
+                Double.parseDouble(value.getString())));
+      }
+    }
+    return list;
+  }
+
+  private List<Pair<Long, Double>> filterData(
+      List<Pair<Long, Double>> dataList, long startTime, long endTime) {
+    return dataList.stream()
+        .filter(v -> v.getKey() >= startTime && v.getKey() <= endTime)
+        .collect(Collectors.toList());
+  }
+
+  MetricTimeSeries getMetricTimeSeriesRecord(MetricSelection metricSelection) {
+    return metricCache.getIfPresent(metricSelection);
+  }
+
+  @Getter
+  @Setter
+  static class MetricTimeSeries {
```

Comment:
I think we should make the TimeSeries a first-class object. This one represents its value object.
We can discuss this offline and can take that up as a follow-up.